File: ant-sdk/antd-cpp/examples/04-files.cpp
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sstream>
#include "antd/antd.hpp"

namespace fs = std::filesystem;

static std::string read_file(const fs::path& path) {
    std::ifstream f(path);
    std::stringstream ss;
    ss << f.rdbuf();
    return ss.str();
}

static void write_file(const fs::path& path, const std::string& content) {
    std::ofstream f(path);
    f << content;
}

int main() {
    fs::path tmp = fs::temp_directory_path() / "antd-cpp-04-files";
    std::error_code ec;
    fs::remove_all(tmp, ec);
    fs::create_directories(tmp);

    const std::string file_content = "Hello from a file on Autonomi!";

    fs::path src_file = tmp / "hello.txt";
    write_file(src_file, file_content);

    try {
        antd::Client client;

        auto est = client.file_cost(src_file.string(), true);
        std::cout << "Estimate: " << est.file_size << " bytes in " << est.chunk_count
                  << " chunks, storage " << est.cost << " atto, gas "
                  << est.estimated_gas_cost_wei << " wei, mode " << est.payment_mode << "\n";

        auto result = client.file_put_public(src_file.string());
        std::cout << "File uploaded at: " << result.address << "\n";
        std::cout << "  storage: " << result.storage_cost_atto << " atto, gas: "
                  << result.gas_cost_wei << " wei\n";
        std::cout << "  chunks: " << result.chunks_stored << ", mode: "
                  << result.payment_mode_used << "\n";

        fs::path dst_file = tmp / "hello.txt.downloaded";
        client.file_get_public(result.address, dst_file.string());
        std::cout << "File downloaded to " << dst_file << "\n";

        std::string got = read_file(dst_file);
        if (got != file_content) {
            std::cerr << "Round-trip mismatch: wrote " << file_content.size()
                      << " bytes, read " << got.size() << " bytes\n";
            return 1;
        }

        std::cout << "File upload/download OK!\n";
    } catch (const antd::AntdError& e) {
        std::cerr << "Error: " << e.what() << "\n";
        fs::remove_all(tmp, ec);
        return 1;
    }

    fs::remove_all(tmp, ec);
    return 0;
}

File: ant-sdk/antd-rust/examples/04-files.rs
use antd_client::{Client, PaymentMode, DEFAULT_BASE_URL};
use std::fs;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let client = Client::new(DEFAULT_BASE_URL);

    let tmp = std::env::temp_dir().join("antd-rust-04-files");
    let _ = fs::remove_dir_all(&tmp);
    fs::create_dir_all(&tmp)?;

    let file_content = "Hello from a file on Autonomi!";

    let src_file = tmp.join("hello.txt");
    fs::write(&src_file, file_content)?;

    let est = client
        .file_cost(src_file.to_str().unwrap(), true, PaymentMode::Auto)
        .await?;
    println!(
        "Estimate: {} bytes in {} chunks, storage {} atto, gas {} wei, mode {}",
        est.file_size, est.chunk_count, est.cost, est.estimated_gas_cost_wei, est.payment_mode
    );

    let result = client
        .file_put_public(src_file.to_str().unwrap(), PaymentMode::Auto)
        .await?;
    println!(
        "File uploaded at: {} (storage: {} atto, gas: {} wei, chunks: {}, mode: {})",
        result.address,
        result.storage_cost_atto,
        result.gas_cost_wei,
        result.chunks_stored,
        result.payment_mode_used,
    );

    let dst_file = tmp.join("hello.txt.downloaded");
    client
        .file_get_public(&result.address, dst_file.to_str().unwrap())
        .await?;
    println!("File downloaded to {}", dst_file.display());

    let got = fs::read_to_string(&dst_file)?;
    if got != file_content {
        fs::remove_dir_all(&tmp).ok();
        return Err(format!(
            "round-trip mismatch: wrote {} bytes, read {} bytes",
            file_content.len(),
            got.len()
        )
        .into());
    }

    fs::remove_dir_all(&tmp).ok();
    println!("File upload/download OK!");
    Ok(())
}
